### ISSUE TITLE
Can't build html with footnotes when using i18n

### ISSUE THREAD

**Issue by shimizukawa** [sphinx-doc/sphinx#955]
Attached is a small test project that fails with sphinx v1.1.3.

When I tried to make html, I got an exception:

```
# Sphinx version: 1.1.3
# Python version: 2.7.3
# Docutils version: 0.9.1 release
# Jinja2 version: 2.6
Traceback (most recent call last):
  File "/opt/local/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/site-packages/sphinx/cmdline.py", line 189, in main
    app.build(force_all, filenames)
  File "/opt/local/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/site-packages/sphinx/application.py", line 204, in build
    self.builder.build_update()
  File "/opt/local/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/site-packages/sphinx/builders/__init__.py", line 196, in build_update
    'out of date' &#37; len(to_build))
  File "/opt/local/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/site-packages/sphinx/builders/__init__.py", line 252, in build
    self.write(docnames, list(updated_docnames), method)
  File "/opt/local/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/site-packages/sphinx/builders/__init__.py", line 292, in write
    self.write_doc(docname, doctree)
  File "/opt/local/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/site-packages/sphinx/builders/html.py", line 419, in write_doc
    self.docwriter.write(doctree, destination)
  File "/opt/local/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/site-packages/docutils/writers/__init__.py", line 77, in write
    self.translate()
  File "/opt/local/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/site-packages/sphinx/writers/html.py", line 38, in translate
    self.document.walkabout(visitor)
  File "/opt/local/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/site-packages/docutils/nodes.py", line 173, in walkabout
    if child.walkabout(visitor):
  File "/opt/local/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/site-packages/docutils/nodes.py", line 173, in walkabout
    if child.walkabout(visitor):
  File "/opt/local/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/site-packages/docutils/nodes.py", line 173, in walkabout
    if child.walkabout(visitor):
  File "/opt/local/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/site-packages/docutils/nodes.py", line 165, in walkabout
    visitor.dispatch_visit(self)
  File "/opt/local/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/site-packages/docutils/nodes.py", line 1611, in dispatch_visit
    return method(node)
  File "/opt/local/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/site-packages/docutils/writers/html4css1/__init__.py", line 973, in visit_footnote_reference
    href = '#' + node['refid']
  File "/opt/local/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/site-packages/docutils/nodes.py", line 541, in __getitem__
    return self.attributes[key]
KeyError: 'refid'
```

When using '[1]_' in index.rst instead of '[#]_', build succeeds.

---
- Bitbucket: https://bitbucket.org/birkenfeld/sphinx/issue/955
- Originally reported by: Arihiro TAKASE
- Originally created at: 2012-06-26T17:25:31.008

**Comment by shimizukawa** [comment]
_From Christophe Simonis on 2012-10-07 20:27:27+00:00_

The localisation is done via a `Transformer` that create a new document with the translation done, then update the initial node with the translated children. But it also copy non textual node, leading to inconsistant document having reference to the old nodes.

This patch solve the problem.

**Comment by shimizukawa** [comment]
_From Christophe Simonis on 2012-10-09 21:55:46+00:00_

Pull request created.

**Comment by shimizukawa** [comment]
_From Georg Brandl on 2012-10-28 16:55:22+00:00_

Close #955: Only update text nodes in translation

→ <<cset 23a86297c345df95f0ea0a2db7e23631514af2ab>>

**Comment by shimizukawa** [comment]
_From Dmitry Shachnev on 2012-10-29 09:08:15+00:00_

Quoting from [this message](https://bugs.launchpad.net/ubuntu-packaging-guide/+bug/1068493/comments/20):

> However there is a real problem on how Sphinx handles the translation. The problem arise whenever a segment to be translated starts or with a link and the translated version adds some text before or after the starting/ending link, respectively. For example:
> - if "`Link`_ English text" is translated into "Spanish text `Link`_ more text" it breaks Sphinx.
> - if "English text `Link`_ " is translated into "Spanish text `Link`_ more text" it breaks Sphinx.
> 
> This should be accepted by the tool, as those are perfectly normal situations (in fact, my modified translation to avoid the problem sounds less natural to me than original ones).

I understand that such cases weren't always working before this patch was applied, but this patch explicitly breaks these cases.

Can you please fix it?

**Comment by shimizukawa** [comment]
_From Robert Lehmann on 2012-11-06 10:06:55+00:00_

I'm going to look into that.

**Comment by shimizukawa** [comment]
_From Dmitry Shachnev on 2012-11-19 10:46:31+00:00_

[Robert Lehmann](https://bitbucket.org/lehmannro): are you still planning to look into this?

**Comment by shimizukawa** [comment]
_From Robert Lehmann on 2012-11-19 12:34:56+00:00_

Absolutely, but I didn't come around to do it yet.  If you feel like digging into it:  please go ahead!

**Comment by shimizukawa** [comment]
_From Takayuki Shimizukawa on 2012-11-22 02:53:58+00:00_

I create a [PullRequest#86](https://bitbucket.org/birkenfeld/sphinx/pull-request/86). Please review and merge it.

**Comment by shimizukawa** [comment]
_From Takayuki Shimizukawa on 2012-11-27 11:11:07+00:00_

related issue #1044 was filed and fixed it.
I updated my pull request #86

**Comment by shimizukawa** [comment]
_From Takayuki Shimizukawa on 2012-12-07 08:24:48+00:00_

fix #1044 refs #955: fix external/inline links with localization

→ <<cset c347b50330a18a9328c62b30aba3892de4a755db>>

**Comment by shimizukawa** [comment]
_From Takayuki Shimizukawa on 2012-12-07 08:24:48+00:00_

fix #955: consider the numbered-footnote. update test and code

→ <<cset f50091b643edb3dc1eeb238fa2110c52e0244c59>>

**Comment by shimizukawa** [comment]
_From Takayuki Shimizukawa on 2012-12-07 08:24:48+00:00_

fixed #955: footnote i18n translation cause KeyError and 'Only update text nodes in translation' change at b7b808e46851 break translation.

→ <<cset 30c3fff6d79a63b4e892aecf5aa84530c48f51a6>>
